CopyToClipboard unit tests

Ribas160 opened the pull request:
This PR is a continuation of #1475

## Changes
* Unit tests for CopyToClipboard js class added

I need an advice from the experienced developer and I would like to ask a few questions if you don't mind :)

In "copy with button click" test I also used `PasteViewer` class
```js
jsc.property('Copy with button click', common.jscFormats(), 'nestring', function (format, text) {
    var clean = jsdom();

    $('body').html(
        '<div id="placeholder" class="hidden">+++ no paste text ' +
        '+++</div><div id="prettymessage" class="hidden">' +
        '<button type="button" id="prettyMessageCopyBtn"><svg id="copyIcon"></svg>' +
        '<svg id="copySuccessIcon"></svg></button><pre ' +
        'id="prettyprint" class="prettyprint linenums:1"></pre>' +
        '</div><div id="plaintext" class="hidden"></div>'
    );

    $.PrivateBin.PasteViewer.init();
    $.PrivateBin.PasteViewer.setFormat(format);
    $.PrivateBin.PasteViewer.setText(text);
    $.PrivateBin.PasteViewer.run();

    $.PrivateBin.CopyToClipboard.init();
    $.PrivateBin.CopyToClipboard.enableTestMode();
    
    $('#prettyMessageCopyBtn').trigger('click');
    
    const copiedText = $.PrivateBin.CopyToClipboard.readFromClipboard();

    clean();

    return text === copiedText;
});
```

I did it, because `CopyToClipboard` class gets paste text from `PasteViewer` class
```js
function handleCopyButtonClick() {
    $(copyButton).click(function() {
        const text = PasteViewer.getText();
        saveToClipboard(text);

        toggleSuccessIcon();
        showAlertMessage('Paste copied to clipboard');
    });
};
```

Due to this fact I think that this test is not purely unit, because it depends on external functional. How do you think is it a bad design from me and I should redesign `CopyToClipboard` class for avoiding this dependency or in this case it is ok?

Another question is about possible refactoring:
I found that we show and hide elements by adding and removing `hidden` class
```js
$emailLink.addClass('hidden');
$cloneButton.addClass('hidden');
$customAttachment.removeClass('hidden');
$fileWrap.removeClass('hidden');
```

What do you think about adding an abstraction level to this functional?
For example, create methods in `UiHelper` class and use them across the app?
```js
showElement = function (element) {
  element.removeClass('hidden');
}

hideElement = function (element) {
  element.addClass('hidden');
}

UiHelper.showElement($emailLink);
UiHelper.hideElement($fileWrap);

```

elrido commented:
Yes, unit testing can get a bit philosophical. For any such helper functions used by unit tests only, I would place these in the exports in `js/common.js`, so they become available to all tests through the common object:
https://github.com/PrivateBin/PrivateBin/blob/42b7b9060dd150c9f89760ff8598bc7cd17a295e/js/common.js#L81

For mocks of functions not provided by jsDOM, if it is possible to do that, I would try and define the two missing functions to read from and write into the clipboard, similar to this one:
https://github.com/PrivateBin/PrivateBin/blob/42b7b9060dd150c9f89760ff8598bc7cd17a295e/js/test/AttachmentViewer.js#L35-L44

I'd prefer this over having to add mock functions or test modes into the main source code - I'm always afraid those might get triggered in normal usage somehow or otherwise introduce unnecessary complexity. But if that is not possible, then so be it - see for example this disabled part of a test where we are waiting on a missing DOM function:
https://github.com/PrivateBin/PrivateBin/blob/42b7b9060dd150c9f89760ff8598bc7cd17a295e/js/test/Helper.js#L61-L68

Ribas160 commented:
@elrido 
I succeed with creating mock functions for `naviator.clipboard`


About helpers: I don't think I made it clear what I meant, sorry.
I mean to add helper functions to the main code in `privatebin.js` and use them there

We call `addClass('hidden')`, `removeClass('hidden)` functions a lot of times there, just one example
https://github.com/PrivateBin/PrivateBin/blob/863c5b320e801178969b5bc64832c7a6f9aeacce/js/privatebin.js#L2767C1-L2773C14
I think a helper will make this code much more readable

```
if (format === 'markdown') {
    UiHelper.showElement($plainText);
    UiHelper.hideElement($prettyMessage);
} else {
    UiHelper.hideElement($plainText);
    UiHelper.showElement($prettyMessage);
}
```

rugk commented:
BTW, [you can also mark code snippets with the language, which adds syntax highlighting to them, in the Markdown you write on GitHub](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks). I have added these for you in the cases above.
Or copy the links to the lines from the original source code on GitHub.com, then you additionally have a reference to those, as you did in your second-last example.

rugk commented:
> How do you think is it a bad design from me and I should redesign CopyToClipboard class for avoiding this dependency or in this case it is ok?

I would not say the code needs refactoring. But in an ideal world, we would mock the PasteViewer module so you can directly return the values that you want without having to mock that all-over again.
This may not be possible here though, so just go with what works.


> I mean to add helper functions to the main code in privatebin.js and use them there
>
> We call addClass('hidden'), removeClass('hidden) functions a lot of times there, just one example

Well… if writing tests turns out you can optimise the code and it improves readability, then maybe do it for sure.
In this particular example I am not sure whether wrapping such a simple call makes sense.

[As for the code you quote](https://github.com/PrivateBin/PrivateBin/blob/863c5b320e801178969b5bc64832c7a6f9aeacce/js/privatebin.js#L2767-L2773) (with if markdown statement etc.) I can only find it once in `privatebin.js` , so judging from that, it is not necessary?  If you would use it somewhere, then yeah, making it a wrapper function would be good.

Ribas160 commented:
@rugk 

> BTW, [you can also mark code snippets with the language, which adds syntax highlighting to them, in the Markdown you write on GitHub](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks). I have added these for you in the cases above.

Thank you for the advice, I didn't know that. I definetely will use it.

> Or copy the links to the lines from the original source code on GitHub.com, then you additionally have a reference to those, as you did in your second-last example.

Yes, I found this feature today :)

> Well… if writing tests turns out you can optimise the code and it improves readability, then maybe do it for sure.
> In this particular example I am not sure whether wrapping such a simple call makes sense.
> 
> [As for the code you quote](https://github.com/PrivateBin/PrivateBin/blob/863c5b320e801178969b5bc64832c7a6f9aeacce/js/privatebin.js#L2767-L2773) (with if markdown statement etc.) I can only find it once in privatebin.js , so judging from that, it is not necessary? If you would use it somewhere, then yeah, making it a wrapper function would be good.

I added the statement just for example, generally we have a lot of places where we hide and show elements by removing and adding the `hidden` class.

elrido commented:
> > I mean to add helper functions to the main code in privatebin.js and use them there
> > We call addClass('hidden'), removeClass('hidden) functions a lot of times there, just one example
> 
> Well… if writing tests turns out you can optimise the code and it improves readability, then maybe do it for sure. In this particular example I am not sure whether wrapping such a simple call makes sense.

Ok, sorry for misunderstanding which code you meant.

For readability, we do already often, but not always, provide such functions, but always specific to the objects in question, for example, in the Editor class:
https://github.com/PrivateBin/PrivateBin/blob/863c5b320e801178969b5bc64832c7a6f9aeacce/js/privatebin.js#L2591-L2593
https://github.com/PrivateBin/PrivateBin/blob/863c5b320e801178969b5bc64832c7a6f9aeacce/js/privatebin.js#L2604-L2606

I tend to agree that for such a simple function, a helper function seem a little bit overkill.

Though, on the other hand, we had discussed [moving away from using jQuery](https://github.com/PrivateBin/PrivateBin/issues/198) and such helper functions would provide a single spot to replace this one use of jQuery methods. It seems we seem to mostly add or remove the class "hidden", but "active" seems also used a lot.

We could offer a set of functions to hide & unhide, which take the element as the first required parameter and a second optional argument for the class name, which is set to "hidden" by default?

elrido reviewed (approved):
I'm more than happy with these changes - Thank you very much!

Is it Ok to merge it?

Ribas160 commented:
@elrido 
> It seems we seem to mostly add or remove the class "hidden", but "active" seems also used a lot.
> 
> We could offer a set of functions to hide & unhide, which take the element as the first required parameter and a second optional argument for the class name, which is set to "hidden" by default?

We use "active" class to make element "selected" not for hiding it. My idea was not just wrap remove/add class functions, but create hide/show element methods. Following the logic we should create active/inactive functions as well for "active" class.

> Though, on the other hand, we had discussed [moving away from using jQuery](https://github.com/PrivateBin/PrivateBin/issues/198) and such helper functions would provide a single spot to replace this one use of jQuery methods.

Don't you want to think about moving to SPA model: use React for example? It could help us to use components approach and we still will be able to use bootstrap5 classes or even [react-bootstrap](https://react-bootstrap.netlify.app/) library.

> Is it Ok to merge it?

Yes, PR is ready to merge

elrido commented:
Ok, adding such methods to the elements themselves sounds very sensible. And yes, it essentially already is an SPA, just currently not using any framework outside of jQuery. I'm not fundamentally opposed to exploring other frameworks, especially if they would let us reduce code on our end. Lets see what other maintainers have to say on this - maybe we should move that discussion into a separate issue or continue discussing in the jQuery removal one?

rugk commented:
Yes, it's a totally different topic and as such off-topic, but for such a relatively simple application I would rather oppose a whole framework like React or Angular. Let's focus on modernizing the code first and removing jQuery, see #198.
